lib/wallaroo/core/sink/connector_sink/connector_sink_notify.pony

```diff
@@ -0,0 +1,455 @@
+              // 3. The connector sink decides to send an ACK with p_o_r=4000.
+              //    This message is delayed just a little bit to make a race.
+              // 4. We process prepare_to_rollback & rollback.  Our
+              //    message_id is reset to message_id=0.  Wallaroo
+              //    has fully rolled back state and is ready to resume
+              //    all of its work from offset=0.
+              // 5. The ACK from step #3 arrives.
+              //    4000 > 0, which looks like a terrible state:
+              //    we haven't sent anything, but the sink is ACKing
+              //    4000 bytes.
+              //
+              // I am going to disable unsolicited ACK'ing by the
+              // connector sink.  Disabling unsolicited ACKs would
+              // definitely prevent late-arriving ACKs.  (Will still
+              // send ACK in response to Eos.)
+              // TODO: Otherwise I believe we'd have to
+              // add a checkpoint #/rollback #/state-something to
+              // the ACK message to be able to identify late-arriving
+              // ACKs.
+              None
+            elseif p_o_r < acked_point_of_ref then
+              @printf[I32]("Error: Ack: stream-id %lu p_o_r %lu acked_point_of_ref %lu\n".cstring(), _stream_id, p_o_r, acked_point_of_ref)
+              Fail()
+            else
+              acked_point_of_ref = p_o_r
+            end
+          else
+            @printf[I32]("Ack: unknown stream_id %d\n".cstring(), s_id)
+            Fail()
+          end
+        end
+      else
+        _error_and_close(conn, "Bad FSM State: F" + _fsm_state().string())
+      end
+    | let m: cp.RestartMsg =>
+      ifdef "trace" then
+        @printf[I32]("TRACE: got restart message, closing connection\n".cstring())
+      end
+      conn.close()
+    end
+
+  fun ref process_uncommitted_list(conn: ConnectorSink ref) =>
+    """
+    In case of a Wallaroo failure, we need to do two things that can
+    happen in either order: 1. get list of uncommitted transactions,
+    2. get the initial rollback() message + payload blob of state.
+    After both have happened, then we need to commit/abort any
+    uncommitted txns outstanding at the connector sink.
+    """
+
+    match (twopc_txn_id_last_committed, twopc_uncommitted_list)
+    | (let last_committed: String, let uncommitted: Array[String] val) =>
+      ifdef "checkpoint_trace" then
+        @printf[I32]("2PC: process_uncommitted_list processing %d items, last_committed = %s\n".cstring(), uncommitted.size(), last_committed.cstring())
+      end
+      for txn_id in uncommitted.values() do
+        let do_commit = if txn_id == last_committed then true else false end
+        ifdef "checkpoint_trace" then
+          @printf[I32]("2PC: uncommitted txn_id %s commit=%s\n".cstring(), txn_id.cstring(), do_commit.string().cstring())
+        end
+        let abort = cp.TwoPCEncode.phase2(txn_id, do_commit)
```

Comment:
Can this be a commit message?  If so, `abort` is misleading

Reply:
Ja, I'll change the var names.